### PR TITLE
ci: use gotestsum for running tests

### PR DESCRIPTION
**Pull request by notnoop**
`gotestsum` has user friendlier output that emits final summary, also it can emit junit xml file for
automated analysis instead of current format that should significantly
ease automated analysis of CI.

The output of gotestsum is nicer for few reasons (a sample output in https://travis-ci.org/hashicorp/nomad/jobs/468476271 ):

* It emits the final summary line, making it easier to identify unexpected things
``
DONE 2949 tests, 39 skipped, 3 failures in 1034.302s
```

* It emits the package of the failing tests along side the test name
```
=== FAIL: drivers/rawexec TestRawExecDriver_StartWaitRecoverWaitStop (0.40s)
=== PAUSE TestRawExecDriver_StartWaitRecoverWaitStop
=== CONT  TestRawExecDriver_StartWaitRecoverWaitStop
2018-12-15T19:30:38.527Z [DEBUG] plugin: starting plugin: path=/home/travis/gopath/bin/nomad args="[/home/travis/gopath/bin/nomad executor {"LogFile":"/tmp/nomad_driver_harness-820714166/sleep/executor.out","LogLevel":"debug","FSIsolation":false}]"
2018-12-15T19:30:38.527Z [DEBUG] plugin: waiting for RPC address: path=/home/travis/gopath/bin/nomad
2018-12-15T19:30:38.582Z [DEBUG] plugin.nomad: plugin address: address=/tmp/plugin181850657 network=unix timestamp=2018-12-15T19:30:38.581Z
    testlog.go:32: 2018-12-15T19:30:38.527Z [INFO ] rawexec/driver.go:309: raw_exec: starting task: driver_cfg="{Command:/tmp/go-build099672588/b827/rawexec.test Args:[sleep 100s]}"
    driver_test.go:303: 
        	Error Trace:	driver_test.go:303
        	Error:      	Received unexpected error:
        	            	cannot destroy running task
        	Test:       	TestRawExecDriver_StartWaitRecoverWaitStop
```

* It emits the skipped tests - which seems in many cases is actually a bug, e.g.

```
=== Skipped
=== SKIP: client TestAllocations_GarbageCollect (0.00s)
    alloc_endpoint_test.go:80: missing mock driver plugin implementation
```

### CODE REVIEW

**Comment by jippi** [comment]
@notnoop I'm super happy with https://github.com/tebeka/go2xunit for jUnit output from `go test` 

```
go get github.com/tebeka/go2xunit
2>&1 go test -v -covermode=atomic -coverprofile "cover/coverage.cov" "$package" ./... | tee cover/out.txt
go2xunit -fail -input cover/out.txt -output cover/junit.xml
```

**Comment by notnoop** [comment]
Thanks, @jippi - I used a similar tool (or was it that one actually, don't remember) in previous life, but we abandoned it then, in favor of gotestsum because we hit few edge cases that made piping pattern unreliable.  We had some cases where the tool failed to detect some failures and exited with zero, despite the original `go test` command exiting with non-zero exit code.

Gotestsum provides quite resilient in these cases, as it relies upon go test json events [1], and outputs useful metadata (e.g. timing information, etc) without risk missing go test exit code.  Combine that with the user-friendly output, it seemed like a winner - and it's currently my go test runner of choice.

[1] https://github.com/golang/go/issues/2981

File: .travis.yml [hunk 1]
```diff
@@ -58,4 +58,7 @@ install:
   - if [[ "$RUN_UI_TESTS" ]]; then . $HOME/.nvm/nvm.sh && cd ui && nvm use && cd .. ; fi
 
 script:
-  - sudo -E "PATH=$PATH" make travis
+  - sudo -E "PATH=$PATH" GOTESTSUM_JUNITFILE=/tmp/results.xml make travis
+
+after_script:
+  - cat /tmp/results.xml
```
Comment (schmichael) [inline .travis.yml]: Do we need to do this right now? Is it useful enough that people will copy and paste it out of Travis?

Comment (notnoop) [reply .travis.yml]: Yes!  At least for me ;-).  I use travis cli[1] to fetch logs and then analyze them locally to identify flaky tests.  When viewed on the TravisCI site, the output is collapsed by default, so hopefully doesn't impact others.

[1] https://github.com/travis-ci/travis.rb

Comment (dantoml) [reply .travis.yml]: +1 from me on this being super useful. Honestly probably my most missed CCI feature is the test results handling.

Comment (schmichael) [reply .travis.yml]: Cool! Very excited about new tools to investigate builds!

**Comment by dadgar** [comment]
This looks great!
